models/db/common.go

```diff
@@ -20,3 +29,337 @@ func BuildCaseInsensitiveLike(key, value string) builder.Cond {
 	}
 	return builder.Like{"UPPER(" + key + ")", strings.ToUpper(value)}
 }
+
+// InsertOnConflictDoNothing will attempt to insert the provided bean but if there is a conflict it will not error out
+// This function will update the ID of the provided bean if there is an insertion
+// This does not do all of the conversions that xorm would do automatically but it does quite a number of them
+// once xorm has a working InsertOnConflictDoNothing this function could be removed.
+func InsertOnConflictDoNothing(ctx context.Context, bean interface{}) (bool, error) {
+	e := GetEngine(ctx)
+
+	tableName := x.TableName(bean, true)
+	table, err := x.TableInfo(bean)
+	if err != nil {
+		return false, err
+	}
+
+	autoIncrCol := table.AutoIncrColumn()
+
+	cols := table.Columns()
+
+	colNames, args, emptyColNames, emptyArgs, err := getColNamesAndArgsFromBean(bean, cols)
+	if err != nil {
+		return false, err
+	}
+
+	if len(colNames) == 0 {
+		return false, fmt.Errorf("provided bean to insert has all empty values")
+	}
+
+	// MSSQL needs to separately pass in the columns with the unique constraint and we need to
+	// include empty columns which are in the constraint in the insert for other dbs
+	uniqueCols, uniqueArgs, colNames, args := addInUniqueCols(colNames, args, emptyColNames, emptyArgs, table)
+	if len(uniqueCols) == 0 {
+		return false, fmt.Errorf("provided bean has no unique constraints")
+	}
+
+	sb := &strings.Builder{}
+	switch {
+	case setting.Database.UseSQLite3 || setting.Database.UsePostgreSQL || setting.Database.UseMySQL:
+		_, _ = sb.WriteString("INSERT ")
+		if setting.Database.UseMySQL && autoIncrCol == nil {
+			_, _ = sb.WriteString("IGNORE ")
+		}
+		_, _ = sb.WriteString("INTO ")
+		_, _ = sb.WriteString(x.Dialect().Quoter().Quote(tableName))
+		_, _ = sb.WriteString(" (")
+		_, _ = sb.WriteString(colNames[0])
+		for _, colName := range colNames[1:] {
+			_, _ = sb.WriteString(",")
+			_, _ = sb.WriteString(colName)
```

Comment:
I am not sure whether the `colName` is already quoted?

Reply:
Now that I've started quoting everything mssql has broken...

Reply:
I can see a failure in https://drone.gitea.io/go-gitea/gitea/67001/2/17, which is "Error inserting package: error in query: %!s(<nil>), Unsupported type error", it doesn't look like to be related to the column name quoting.

And I really think this new function needs its own test cases, instead of depending on the package tests.